### DISCUSSION TITLE
Vinculação de Dados a Empresas em uma Aplicação SaaS

### DISCUSSION THREAD

**Discussion by reluiz** [balta-io/forum#573] [Tira Dúvidas] [answered]
Estou gostando muito do curso e do conteúdo sobre como desenvolver uma aplicação SaaS com dados vinculados a um usuário. Tenho uma dúvida sobre um cenário diferente.

Como eu poderia adaptar a estrutura da aplicação para vincular os dados a uma empresa, considerando que uma empresa pode ter vários usuários? Quais seriam as melhores práticas para implementar isso em termos de modelagem de dados e controle de acesso?

Agradeço desde já pela ajuda.

Atenciosamente,
Renato Luiz

**Comment by andrebaltieri** [comment, marked as answer]
Boa tarde, @reluiz, como vai?

Muito obrigado pelo feedback, fico feliz que esteja gostando.
Sobre o cenário que citou, tem como sim, teria que criar uma entidade para empresa e vincular o usuário a ela.
Ai os dados podem ser ou do usuário ou da empresa.

Sobre as melhores práticas, depende muito do cenário... pode ser algo bem simples ou extremamente complexo.
De qualquer forma, a estrutura é um começo!

**Reply by maojsm** [reply]
O Identity tem algum suporte pra  isso (multi-tenant)?

**Comment by reluiz** [comment]
Boa tarde, @andrebaltieri!

Você poderia me dizer se já gravou ou publicou algum material que eu pudesse consultar e usar como referência?

**Reply by maojsm** [reply]
Estou com o mesmo "problema", seria ótimo ter um material sobre isso usando o Identity.

**Reply by andrebaltieri** [reply]
Não me recordo mas a ideia é bem simples...
- Fazer uma requisição de login na API
- Receber o Token
- Armazenar o Token no local storage
- A cada requisição enviar o token no cabeçalho

Posso colocar no pipe aqui para produzir algo sobre

**Comment by reluiz** [comment]
@andrebaltieri Boa tarde! Gostaria de pedir sua ajuda para resolver essa dúvida. Estou estudando bastante os cursos do balta, mas mesmo assim não consegui compreender como eu posso preparar a aplicação que estou trabalhando para atender empresas que tenham 1(um) ou mais usuários.

**Reply by andrebaltieri** [reply]
Bom dia, @reluiz, como vai? Claro... vamos lá...
Quais problemas está tendo e o que já tentou como solução? Consegue esboçar algo aqui e eu valido se está OK?

**Reply by reluiz** [reply]
> Bom dia, @reluiz, como vai? Claro... vamos lá... Quais problemas está tendo e o que já tentou como solução? Consegue esboçar algo aqui e eu valido se está OK?

Bom dia, @andrebaltieri! Vou tentar explicar aqui as minhas dúvidas. 

Bom primeiro gostaria de expor o contexto para ajudar no entendimento. Estou desenvolvendo um SaaS um (MVP) na verdade, mas será um SaaS para atender micro e pequenas empresas. Essas empresas poderão ter um ou vários usuários e aí começou a minha dúvida. No curso FullStack você nos ensina como podemos recuperar dos Claims o usuário que está logado e com essa informação conseguimos salvar por exemplo uma categoria vinculando o usuário a essa categoria.

# O que tentei fazer?

Primeiro criei uma tabela chamada Company

Depois nas tabelas como Category e nas outras eu adicionei um novo campo chamado CompanyId e gerei as migrações para o 
EF criar os relacionamentos.

Já no código eu movi para o projeto Core onde modifiquei a classe Request e adicionei uma a propriedade
```csharp
public string CompanyId {get; set;}
```

e modifiquei também a classe Account adicionando também a propriedade
```csharp
public string CompanyId {get; set;}
```

Em segui alterei a classe CookieAuthenticationStateProvider no metodo GetClaims() adicionando um novo Claim chamado "CompanyID e no valor deixei fixo "1" para testar. 

```csharp
 var claims = new List<Claim>
 {
     new(ClaimTypes.Name, user.Email),
     new(ClaimTypes.Email, user.Email),
     new("CompanyId", "1"),
 };

 claims.AddRange(
     user.Claims.Where(x =>
             x.Key != ClaimTypes.Name &&
             x.Key != ClaimTypes.Email)
         .Select(x
             => new Claim(x.Key, x.Value))
 );
```

Após isso criei no projeto Api um diretório chamado Extensions e adicionei o método de extensão para extrair esse Cliam "CompanyId"

```csharp
namespace Flant.Api.Extensions
{
    public static class ClaimsPrincipalExtension
    {
        public static int CompanyId(this ClaimsPrincipal claims)
        {
            try
            {
                var value = claims?.FindFirst("CompanyId")?.Value;
                return int.Parse(value);
            }
            catch
            {
                return 0;
            }
        }
    }
}
```

Após isso alterei a classe CreateCategoryEndpoint adicionando a linha  request.CompanyId = user.CompanyId() como mostro abaixo, porém o método sempre retorno 0 

```csharp
private static async Task<IResult> HandleAsync(
    ClaimsPrincipal user,
    ICategoryHandler handler,
    CreateCategoryRequest request)
{

    request.CompanyId = user.CompanyId(); 
    request.UserId = user.Identity?.Name ?? string.Empty;

    var result = await handler.CreateAsync(request);
    
    return result.IsSuccess 
        ? TypedResults.Created($"/{result.Data?.Id}", result) 
        : TypedResults.BadRequest(result);
}
```

Esses foram os passos que tentei, espero que tenha conseguido ser claro. Mas, mesmo deixando um valor fixo no Claims o retorno tem sido sempre 0(zero).  

Se esse meu teste tivesse dado certo eu ainda faria o seguinte:

Iria criar na tabela IdentityUser uma nova coluna chamada CompanyId para vincular a empresa ao usuário. Então após o processo de registro que cria o usuário eu iria também criar a empresa e atualizar o usuário adicionando o id da empresa ao usuário para conseguir recuperar o id da empresa pelo Claims.

**Reply by andrebaltieri** [reply]
Bom dia, @reluiz, como vai?

Até a parte do campo no banco, acho que é isto mesmo, agora sobre o Claim, eu não confiaria em deixar esta informação direto ali.
Primeiro por que de qualquer forma você precisaria validar se o usuário pertence realmente aquela empresa, segundo por que o usuário (Aparentemente) só pertence a uma empresa.

Logo, tendo o UserId você já sabe de qual empresa a informação é, ou me enganei?
O seu trabalho seria mais na parte de vincular o usuário a uma empresa...

Me pontua aí se é realmente isto por favor!

**Reply by reluiz** [reply]
> Bom dia, @reluiz, como vai?
> 
> Até a parte do campo no banco, acho que é isto mesmo, agora sobre o Claim, eu não confiaria em deixar esta informação direto ali. Primeiro por que de qualquer forma você precisaria validar se o usuário pertence realmente aquela empresa, segundo por que o usuário (Aparentemente) só pertence a uma empresa.
> 
> Logo, tendo o UserId você já sabe de qual empresa a informação é, ou me enganei? O seu trabalho seria mais na parte de vincular o usuário a uma empresa...
> 
> Me pontua aí se é realmente isto por favor!

Sim! Nesse MVP o usuário pertence a uma empresa, mas a empresa poderia ter mais usuários.

**Reply by andrebaltieri** [reply]
Se for mult-empresas, a implementação inicial que mostrou é satisfatória. Só frisando novamente, não deixa nenhuma informação sensível no Claim! Se for o ID da empresa, tudo bem, mais que isto não.

Em relação a sempre retornar zero, pode ser que esteja dando algum erro, seu `catch` faz retornar zero, chegou a debugar pra ver?

Outro ponto é que sempre que você altera algo no Claim, precisa fazer logout e login novamente.

Tenta estes pontos e me retorna por favor!

**Comment by barbosamj** [comment]
olá @reluiz tudo bem?
Como vc resolveu a questão de ler o CompanyId no Claims? Estou tendo o mesmo problema de retornar valor incorreto.

**Reply by maojsm** [reply]
Eu usei Roles, nao sei se é o correto mas esta me atendendo bem por hora.
